src/main.clj

```diff
@@ -6,57 +6,80 @@
 
 (def ope
   (insta/parser
-    "S = <nl>* (exp|Q) (<nl>+ (exp|Q))* <nl>*
-    exp = A|M|D|Dp|Ap|Mp
+    "S = <nl>* sttmt (<nl>+ sttmt)* <nl>*
+    <sttmt> = exp | assign | Q
+    assign = Aname <W*> <'='> <W*> exp
+    Aname = name
+    exp = A|M|D|Dp|Ap|Mp|name
     <Ap> = <'('> A <')'>
-    A = (D|Dp|M|Mp|Ap) <W*> (<'+'> <W*> (D|Dp|M|Mp|Ap))+
-    M = (D|Dp|Mp|Ap) <W*> (<'*'> <W*> (D|Dp|Ap|Mp))+
     <Mp> = <'('> M <')'>
     <Dp> = <'('> D <')'>
+    <namep> = <'('> name <')'>
+    A = (elem|M|Mp|Ap) <W*> (<'+'> <W*> (elem|M|Mp|Ap))+
+    M = (elem|Mp|Ap) <W*> (<'*'> <W*> (elem|Ap|Mp))+
     D = #'\\d+'
+    <elem> = D | Dp | Rname
+    Rname = name | namep
     W = #' '
     Q = 'quit!'|'q!'
+    <name> = #'[a-zA-Z]\\w*'
     <nl> = '\n'
     "))
 
 (defn opeeval
-  [txt]
-  (let [tree (ope txt)]
-    (insta/transform {:D (fn d [x] (edn/read-string x))
-                      :A (fn a [& r] (apply + r))
-                      :M (fn m [& r] (apply * r))
-                      :exp (fn bla [x] [:result x])
-                      :S (fn s  [& r] r) ;; we keep all the thingy ...
+  [txt m]
+  (let [tree (ope txt)
+        a (atom m)]
+    (insta/transform {:D (fn D [x] (edn/read-string x))
+                      :A (fn A [& r] (apply + r))
+                      :M (fn M [& r] (apply * r))
+                      :exp (fn exp [x] [:result x])
+                      :S (fn S  [& r] r) ;; we keep all the thingy ...
                       ;; which means we are not handling statement, including q! at all.
-                      :Q (fn q  [x] [:stop])}
+                      :assign (fn assign [x y] (if (and (= (first x) :Aname) (= (first y) :result))
+                                                 (do
+                                                   (swap! a assoc (second x) (second y))
+                                                   [:mem (second x) (second y)])
+                                                 [:error :assign]))
+                      :Rname (fn Rname [x] (get @a x))
+                      :Q (fn Q  [x] [:stop])}
                      tree)))
 
 (defn run
   []
   (print "\nWelcome to proglang!\nThe programming language made for fun and learning :-)
          \n\nFor now, only the + and * operations exist.\nHave fun!\n\n")
-  (loop [stop false]
+  (loop [stop false
+         m {}]
     (if stop
       :stop
       (let [_ (print "> ")
             _ (flush)
             inp (read-line)
-            outp (last (opeeval inp))
+            outp (last (opeeval inp m))
             mss (case (first outp)
                   :stop "\nBye!!\n"
                   :result (str " " (second outp) "\n")
+                  :mem ""
```

Comment:
Since you have an interface for good feedback here, it may be useful to print the assignment as something like:

```
varname := computed_value
```

i.e.
```suggestion
                  :mem (format "%s := %s" (second outp) (last outp))
```